### PR TITLE
Support (dual) power cones

### PR DESCRIPTION
**Pull request by dourouc05**
Follows #129 

Thanks @blegat !

### CODE REVIEW

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -134,6 +134,18 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
...
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.PowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
+function _allocate_constraint(cone::ConeData, f, s::MOI.PowerCone)
+    ci = length(cone.p)
+    push!(cone.p, s.exponent)
+    ci
+end
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.DualPowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
+function _allocate_constraint(cone::ConeData, f, s::MOI.DualPowerCone)
+    ci = length(cone.p)
+    push!(cone.p, s.exponent)
```
Comment (blegat) [inline src/MOI_wrapper.jl]: It seems to me that with the current implementation, the `DualPowerCone` will be treated as a `PowerCone`

Comment (dourouc05) [reply src/MOI_wrapper.jl]: You're right, here is the corrected code.

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -134,6 +134,18 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
...
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.PowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
+function _allocate_constraint(cone::ConeData, f, s::MOI.PowerCone)
+    ci = length(cone.p)
+    push!(cone.p, s.exponent)
+    ci
+end
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.DualPowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
+function _allocate_constraint(cone::ConeData, f, s::MOI.DualPowerCone)
+    ci = length(cone.p)
+    push!(cone.p, - s.exponent)
```
Comment (blegat) [inline src/MOI_wrapper.jl]: I didn't know the dual power cone is simply the power cone with opposite exponent. Could you give a reference or an argument to support the claim ?

Comment (dourouc05) [reply src/MOI_wrapper.jl]: This is the convention for SCS. There are several mentions of it in the source code: 
https://github.com/cvxgrp/scs/blob/b03e156d06bec13c3190c47b4151d94bec4a6a55/src/cones.c#L775
https://github.com/cvxgrp/scs/blob/b03e156d06bec13c3190c47b4151d94bec4a6a55/include/scs.h#L73-75
However, I did not find anything about it in the documentation…

Comment (blegat) [reply src/MOI_wrapper.jl]: Oh I see, it is not equivalent but that's how you specify it to SCS. Could you add a comment explaining that ?

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -134,6 +134,19 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
     cone.ep += 1
     ci
 end
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.PowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
+function _allocate_constraint(cone::ConeData, f, s::MOI.PowerCone)
+    ci = length(cone.p)
+    push!(cone.p, s.exponent)
+    ci
```
Comment (blegat) [inline src/MOI_wrapper.jl]: Missing a return to satisfy the [JuMP style guide](http://www.juliaopt.org/JuMP.jl/dev/style/)

File: src/MOI_wrapper.jl [hunk 4]
```diff
@@ -134,6 +134,19 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
     cone.ep += 1
     ci
 end
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.PowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
```
Comment (blegat) [inline src/MOI_wrapper.jl]: Exceeds 80 character limit: see the [JuMP style guide](http://www.juliaopt.org/JuMP.jl/dev/style/)

File: src/MOI_wrapper.jl [hunk 5]
```diff
@@ -134,6 +134,19 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
...
+function _allocate_constraint(cone::ConeData, f, s::MOI.PowerCone)
+    ci = length(cone.p)
+    push!(cone.p, s.exponent)
+    ci
+end
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.DualPowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
```
Comment (blegat) [inline src/MOI_wrapper.jl]: 80 chars

File: src/MOI_wrapper.jl [hunk 6]
```diff
@@ -134,6 +134,19 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
...
+function _allocate_constraint(cone::ConeData, f, s::MOI.DualPowerCone)
+    ci = length(cone.p)
+    # SCS' convention: dual cones have a negative exponent. 
+    push!(cone.p, - s.exponent)
+    ci
```
Comment (blegat) [inline src/MOI_wrapper.jl]: return

File: src/MOI_wrapper.jl [hunk 7]
```diff
@@ -134,6 +134,19 @@ function _allocate_constraint(cone::ConeData, f, s::MOI.ExponentialCone)
...
+function _allocate_constraint(cone::ConeData, f, s::MOI.PowerCone)
+    ci = length(cone.p)
+    push!(cone.p, s.exponent)
+    ci
+end
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:MOI.DualPowerCone}) = cone.f + cone.l + cone.q + cone.s + cone.ep + ci.value
+function _allocate_constraint(cone::ConeData, f, s::MOI.DualPowerCone)
+    ci = length(cone.p)
+    # SCS' convention: dual cones have a negative exponent. 
+    push!(cone.p, - s.exponent)
```
Comment (blegat) [inline src/MOI_wrapper.jl]: No space for unary operator : http://www.juliaopt.org/JuMP.jl/dev/style/#Whitespace-1

File: src/MOI_wrapper.jl [hunk 8]
```diff
@@ -98,45 +98,76 @@ end
...
 
 const ZeroCones = Union{MOI.EqualTo, MOI.Zeros}
-const LPCones = Union{MOI.GreaterThan, MOI.LessThan, MOI.Nonnegatives, MOI.Nonpositives}
+const LPCones = Union{MOI.GreaterThan, MOI.LessThan, 
+                      MOI.Nonnegatives, MOI.Nonpositives}
 
 # Computes cone dimensions
-constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:ZeroCones}) = ci.value
-#_allocateconstraint: Allocate indices for the constraint `f`-in-`s` using information in `cone` and then update `cone`
+constroffset(cone::ConeData, ci::CI{<:MOI.AbstractFunction, <:ZeroCones}) =
```
Comment (blegat) [inline src/MOI_wrapper.jl]: Do not break a single line function into multiple lines, do `function constroffset(...) ..` instead

Comment (blegat) [reply src/MOI_wrapper.jl]: It should be written in http://www.juliaopt.org/JuMP.jl/dev/style/#TODO:-Line-breaks-1

**Comment by dourouc05** [comment]
Better this way?

File: src/MOI_wrapper.jl (hunk 8 above)
Comment (dourouc05) [reply src/MOI_wrapper.jl]: By the way, I did not see either the 80-character limit in the style guide.

**Pull request merged**